firmware/apps/nic/pv.uc

```diff
@@ -74,23 +74,27 @@ passert(PV_MAX_CLONES, "EQ", 2)
  *       +-+---+-----+-------------------+-----+-------------------------+
  *    10 |A|    0    |   Packet Number   |  0  |         Offset          | 2
  *       +-+---------+-------------------+-----+---------+---------------+
- *    11 |        Sequence Number        | --- | Seq Ctx |   Protocol    | 3
+ *    11 |        Sequence Number        | XML | Seq Ctx |   Protocol    | 3
  *       +-------------------------------+-+-+-+---------+---+-+-+-+-+-+-+
- *    12 |         TX Host Flags         |M|B|Seek (64B algn)|-|Q|I|i|C|c| 4
+ *    12 |         TX Host Flags         |M|B|Seek (64B algn)|X|Q|I|i|C|c| 4
```

Comment:
These changes will clash with other work Netronome has incoming. Specifically, we want to grow the Seq Ctx field by one bit, and we have a different use for the X bit. We might have to add another word to the PV, which has a knock-on effect everywhere else. @mrapson-netronome and I are discussing this.

Reply:
Sure, I was ready to re-implement the layout when pushing the patch.  For me ideal would be to have at least 4+ bits to store the length of XDP meta (in words).

XDP meta is now <= 32 bytes and in 4 bytes chunks, so I need at least 4 bits to store the length. Maybe in future it will be extended to 64 or 128 bytes (still in 4 bytes chunks though), so I would like to have 6+ bits, so that we won't have to change firmware/kernel ABI for this change.

Reply:
The situation is that we are running out of local memory on the datapath ME's:
This is the layout without your change:

```
512                 i32.me0._PKT_IO_PKT_VEC
256                 i32.me0.gro_cli_lm_ctx
32                  i32.me0.__actions_sriov_keys
16                  i32.me0.nfd_in_ring_info
16                  i32.me0.nfd_out_ring_info
2048                i32.me0.EBPF_STACK_BASE
Total used 2880 of 4096B
```
With your change:
```
1024                i32.me0._PKT_IO_PKT_VEC
256                 i32.me0.gro_cli_lm_ctx
32                  i32.me0.__actions_sriov_keys
16                  i32.me0.nfd_in_ring_info
16                  i32.me0.nfd_out_ring_info
2048                i32.me0.EBPF_STACK_BASE
Total used 3392B of 4096B
```

You will notice that changing PV_SIZE_LW from 16 to 24 doubles the _PKT_IO_PKT_VEC size. That is because it must be a power-of-2 size. See the alloc logic in pv.uc line 304. The upshot is that there is 8 more words available in the PV that is not shown in the ASCII art. Perhaps you can move your extra fields to this unused section?

If we add to the above the features that Netronome plans to introduce soon:
```
1024                i32.me0._PKT_IO_PKT_VEC
640                 i32.me0.gro_cli_lm_ctx
16                  i32.me0.nfd_in_ring_info
16                  i32.me0.nfd_out_ring_info
32                  i32.me0._PKT_IO_RX_META
256                 i32.me0.__actions_keys <- had to "break" alignment
2048                i32.me0.EBPF_STACK_BASE
Total used 4032B of 4096B
```
You can see that we are really out of LM, (and I have to violate the alignment requirement of symbol __actions_keys). Thus there really isn't space to grow the XDP metadata.

 If we need to go above 32B, we'll have to consider alternatives, like maintaining the metadata in CLS, or directly in the packet buffer.